File: apps/consumer/android/wear/src/main/java/com/orderingintelligence/consumer/wear/WatchFuelUtils.kt
package com.orderingintelligence.consumer.wear

import kotlin.math.roundToInt
import org.json.JSONObject

data class WatchFuelPaymentPlan(
  val amountCents: Int,
  val paymentFlow: String,
  val requestedAmountCents: Int,
  val requestedLiters: Double,
  val preauthAmountCents: Int
)

fun resolveFuelPreauthCap(session: WatchSessionInfo?, fallbackCents: Int): Int {
  if (session != null) {
    if (session.fuelPreauthCapCents > 0) return session.fuelPreauthCapCents
    if (session.fuelDefaultPrepayCents > 0) return session.fuelDefaultPrepayCents
  }
  return fallbackCents
}

fun resolveFuelPaymentPlan(
  fuel: WatchFuelOrder,
  preauthCapCents: Int
): WatchFuelPaymentPlan? {
  var paymentFlow = fuel.paymentFlow.ifBlank { "prepay" }.lowercase()
  var requestedAmount = fuel.requestedAmountCents
  val requestedLiters = fuel.requestedLiters
  val unitPrice = fuel.unitPriceCents
  var preauthAmount = fuel.preauthAmountCents

  if (paymentFlow == "preauth" && preauthAmount <= 0) {
    preauthAmount = preauthCapCents
  }
  if (paymentFlow != "preauth" && requestedAmount <= 0 && requestedLiters <= 0) {
    paymentFlow = "preauth"
    preauthAmount = preauthCapCents
  }
  if (paymentFlow == "preauth") {
    if (preauthAmount <= 0) return null
    return WatchFuelPaymentPlan(
      amountCents = preauthAmount,
      paymentFlow = "preauth",
      requestedAmountCents = 0,
      requestedLiters = requestedLiters,
      preauthAmountCents = preauthAmount
    )
  }
  val amountCents = when {
    requestedAmount > 0 -> requestedAmount
    requestedLiters > 0 && unitPrice > 0 -> (requestedLiters * unitPrice).roundToInt()
    else -> return null
  }
  return WatchFuelPaymentPlan(
    amountCents = amountCents,
    paymentFlow = paymentFlow,
    requestedAmountCents = requestedAmount,
    requestedLiters = requestedLiters,
    preauthAmountCents = 0
  )
}

fun buildFuelPayload(
  fuel: WatchFuelOrder,
  plan: WatchFuelPaymentPlan
): JSONObject {
  return JSONObject()
    .put("fuelGradeId", fuel.fuelGradeId)
    .put("fuelGradeName", fuel.fuelGradeName)
    .put("unit", "liter")
    .put("unitPriceCents", fuel.unitPriceCents)
    .put("requestedAmountCents", plan.requestedAmountCents)
    .put("requestedLiters", plan.requestedLiters)
    .put("preauthAmountCents", plan.preauthAmountCents)
    .put("paymentFlow", plan.paymentFlow)
}

File: apps/consumer/ios/WatchApp/WatchFuelUtils.swift
import Foundation

struct WatchFuelPaymentPlan {
  let amountCents: Int
  let paymentFlow: String
  let requestedAmountCents: Int
  let requestedLiters: Double
  let preauthAmountCents: Int
}

func resolveFuelPreauthCap(session: WatchSessionInfo?, fallbackCents: Int) -> Int {
  if let cap = session?.fuelPreauthCapCents, cap > 0 { return cap }
  if let storeCap = session?.fuelDefaultPrepayCents, storeCap > 0 { return storeCap }
  return fallbackCents
}

func resolveFuelPaymentPlan(
  fuel: WatchFuelOrder,
  preauthCapCents: Int
) -> WatchFuelPaymentPlan? {
  var paymentFlow = (fuel.paymentFlow ?? "prepay").lowercased()
  var requestedAmount = fuel.requestedAmountCents ?? 0
  let requestedLiters = fuel.requestedLiters ?? 0
  let unitPrice = fuel.unitPriceCents ?? 0
  var preauthAmount = fuel.preauthAmountCents ?? 0

  if paymentFlow == "preauth" && preauthAmount <= 0 {
    preauthAmount = preauthCapCents
  }
  if paymentFlow != "preauth" && requestedAmount <= 0 && requestedLiters <= 0 {
    paymentFlow = "preauth"
    preauthAmount = preauthCapCents
  }
  if paymentFlow == "preauth" {
    guard preauthAmount > 0 else { return nil }
    return WatchFuelPaymentPlan(
      amountCents: preauthAmount,
      paymentFlow: "preauth",
      requestedAmountCents: 0,
      requestedLiters: requestedLiters,
      preauthAmountCents: preauthAmount
    )
  }
  let amountCents: Int
  if requestedAmount > 0 {
    amountCents = requestedAmount
  } else if requestedLiters > 0 && unitPrice > 0 {
    amountCents = Int((requestedLiters * Double(unitPrice)).rounded())
  } else {
    return nil
  }
  return WatchFuelPaymentPlan(
    amountCents: amountCents,
    paymentFlow: paymentFlow,
    requestedAmountCents: requestedAmount,
    requestedLiters: requestedLiters,
    preauthAmountCents: 0
  )
}

func buildFuelPayload(
  fuel: WatchFuelOrder,
  plan: WatchFuelPaymentPlan
) -> [String: Any] {
  return [
    "fuelGradeId": fuel.fuelGradeId,
    "fuelGradeName": fuel.fuelGradeName ?? "",
    "unit": "liter",
    "unitPriceCents": fuel.unitPriceCents ?? 0,
    "requestedAmountCents": plan.requestedAmountCents,
    "requestedLiters": plan.requestedLiters,
    "preauthAmountCents": plan.preauthAmountCents,
    "paymentFlow": plan.paymentFlow
  ]
}
